transcom/mymove
# INT-B-21506

## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A1018853)

## Summary
This story was created to handle the common error cases when updating the port for a service item. By the time I got around to working this story I realized that all the common error cases were already handle in the upstream story for updating the port (B-21505). This story will still be used for testing these error cases so the testing load on B-21505 isn't too high. Since there was no validation code needed for the port error handling we decided to use this story to update the port locations model to be readonly as this reflects the current business rules which weren't settled on until recently.

Is there anything you would like reviewers to give additional scrutiny?

[this article](tbd) explains more about the approach used.

## Verification Steps for the Author

These are to be checked by the author.


## Verification Steps for Reviewers

These are to be checked by a reviewer.


### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

Note: the functionality for ordering POEFSC and PODFSC service items does not yet exist. Therefore in order to test you will need to manually create either a POEFSC or PODFSC service item for a shipment.

1. Create an INTL UB shipment as a customer (CONUS -> AK or AK -> CONUS)
2. Manually create the mto_service_items table for the shipment: `
INSERT INTO mto_service_items (
					id,
                    mto_shipment_id,
                    move_id,
                    re_service_id,
                    status,
                    created_at,
                    updated_at,
                    approved_at
                )
                VALUES (
                uuid_generate_v4(),
                    '8eeba537-bb40-4f58-a534-74ea9ff64590',
                    '813b7a43-2e3f-42f3-8c4e-eca95390d5ca',
                    '388115e8-abe9-441d-96cf-a39f24baa0a3',
                    'APPROVED'::service_item_status,
                    NOW(),
                    NOW(),
                    NOW()
                );
`
3. In the prime v3 api, call getMoveTaskOrder using the move_id for the service item you created in step 2. You will need the etag for the newly created service item for the next step.
4. In the prime v1 api, call updateMTOServiceItem using a payload like this (adjust for you specific service item ID): `
{
  "id": "40b7e71a-6b03-4413-8601-dea694253ee3",
  "modelType": "UpdateMTOServiceItemInternationalPortFSC",
  "portCode": "ABC",
  "reServiceCode": "PODFSC"
}
`
5. Verify that you receive an error response like this: `
{
  "detail": "Unsupported port code (ABC): No port location found for port code ABC",
  "instance": "cffb4ea1-311b-4842-bdf3-bfad74e6f722",
  "title": "Validation Error",
  "invalidFields": {}
}`
6. In the prime v1 api, call updateMTOServiceItem using a payload like this (if you're service item is for PODFSC, pass in POEFSC and vice versa): `
{
  "id": "af98b57c-f3dc-46c2-b12e-0e42ba221b02",
  "modelType": "UpdateMTOServiceItemInternationalPortFSC",
  "portCode": "PDX",
  "reServiceCode": "POEFSC"
}
`
7. Verify that you receive an error response like this: `
{   
"detail": "ID: af98b57c-f3dc-46c2-b12e-0e42ba221b02 is in a conflicting state POE Location can only be updated for service item POEFSC",   
"instance": "6c023056-221b-46bd-b1bb-c2e8cf0dd231",   
"title": "Conflict Error" }
`
8. In the prime v1 api, call updateMTOServiceItem using a payload like this`
{
  "id": "af98b57c-f3dc-46c2-b12e-0e42ba221b02",
  "modelType": "UpdateMTOServiceItemInternationalPortFSC",
  "portCode": "PDX",
  "reServiceCode": "IUBPK"
}
`
9. Verify that you receive an error response like this: `
{
  "title": "Validation Error",
  "instance": "bd14418a-b11f-4409-be69-26df3fc29c08",
  "detail": "reServiceCode in body should be one of [PODFSC POEFSC]"
}
`

### Frontend


### Backend


### Database

#### Any new migrations/schema changes:


## Screenshots
